enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,491 @@
+	// connections).  Headers cannot be set for TLS passthrough
+	// connections.
+	//
+	// +kubebuilder:validation:Optional
+	// +unionMember
+	Set *IngressControllerSetHTTPHeader `json:"set,omitempty"`
+}
+```
+
+```go
+// IngressControllerHTTPHeaderAction defines operation you can perform on headers.
+type IngressControllerHTTPHeaderAction string
+
+const (
+	// Set is used specify set action per header
+    Set    IngressControllerHTTPHeaderAction = "Set"
+	// Delete is used specify delete action per header
+    Delete IngressControllerHTTPHeaderAction = "Delete"
+)
+```
+
+The IngressControllerSetHTTPHeader type has fields for specifying a header value.
+```go
+// IngressControllerSetHTTPHeader defines the value which needs to be set on an HTTP header
+type IngressControllerSetHTTPHeader struct {
+    // Value specifies a header value.
+    // +kubebuilder:validation:Required
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=4096
+    Value string `json:"value"`
+}
+```
+
+B) Set/Delete HTTP response headers in the backend section (plain http backend or backend with TLS terminated at the edge or a
+secure backend with re-encryption) of the haproxy.config via Route's Spec.
+The Route API's Spec is extended by adding optional `HTTPHeaders` field of type `RouteHTTPHeaders` :
+```go
+type RouteSpec struct {
+	...
+
+	// httpHeaders defines policy for HTTP headers.
+	//
+	// If this field is empty, the default values are used.
+	//
+	// kubebuilder:validation:Optional
+	HTTPHeaders *RouteHTTPHeaders `json:"httpHeaders,omitempty" protobuf:"bytes,9,opt,name=httpHeaders"`
+}	
+
+```
+
+```go
+// RouteHTTPHeaders defines policy for HTTP headers.
+type RouteHTTPHeaders struct {
+    // headersManipulation specifies options for set/delete headers.
+	// +kubebuilder:validation:Optional
+	HeadersManipulation RouteHTTPHeadersManipulation `json:"headersManipulation,omitempty" protobuf:"bytes,1,opt,name=headersManipulation"`
+}
+```
+
+```go
+// RouteHTTPHeadersManipulation defines whether we want to set Response or Request Headers.
+// Currently set/delete of HTTP Response Headers is available only.
+type RouteHTTPHeadersManipulation struct {
+    // response is a list of HTTP Headers which either need to set or deleted.
+    Response []RouteSetOrDeleteHTTPHeader `json:"response,omitempty" protobuf:"bytes,1,opt,name=response"`
+}
+```
+
+```go
+// RouteSetOrDeleteHTTPHeader specifies options for set/delete a header.
+type RouteSetOrDeleteHTTPHeader struct {
+    // name specifies a header name to be set or deleted.  Its value must be a valid HTTP header.
+    // name as defined in RFC 2616 section 4.2.
+    // +kubebuilder:validation:Required
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=256
+    // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"
+    Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
```

Comment:
What would happen if I were to create multiple/conflicting entries with the same name in the list of headers?

Reply:
The latest value of the header in the list is set.

Reply:
You need to make the slice for `request` and `response` into a `+listType=map` with `+listMapKey=name` to prevent duplicate entries with the same name, that will be less confusing for users since the API server will only persist a single value for each map key

Reply:
api has been updated.